### ISSUE TITLE
Porting to stm32f103

### ISSUE THREAD

**Issue by Laczen** [Traumflug/Teacup_Firmware#313]
Hi, I would like to port Teacup to a stm32f103 processor (blue pill). I am not experienced with the stm32 devices. Has anyone already ported Teacup?
If it has not been ported yet, what are the steps required? Based on the stm32f411 that is supported, I think that I would need to add cmsis files, but where can I find these?

Thanks and kind regards,

Jehudi

**Comment by Wurstnase** [comment]
Yes this is also ok for a start.
This file https://github.com/stm32duino/Arduino_Core_STM32/blob/master/system/Drivers/CMSIS/Device/ST/STM32F1xx/Include/stm32f103x6.h is the most important.

**Comment by ryannining** [comment]
you can try look at my firmware, some teacup code live in there (gcode parser, serial log) but with different approach on motion planning, to accomodate even without using hardware timer.

Currently its support bluepill and wemos esp8266 for CNC and 3D printer.

https://github.com/ryannining/karyacontroller

**Comment by Laczen** [comment]
Hi @ryannining, thanks for pointing me in the direction of karyacontroller. It seems you have done a lot of work on this. I guess your code still needs some cleaning up. I have seen that you had some issues with speed, has this been resolved ?

**Comment by Laczen** [comment]
@Wurstnase, I have been able to get the cmsis-files in the source tree, I have created a new board, and modified the makefile. I have included the changes in a zip file below.
[stm32f103.zip](https://github.com/Traumflug/Teacup_Firmware/files/2558486/stm32f103.zip)

I know I will have to make changes to several files but I first tried to make and I got the following errors:
  LINK      build/teacup.elf
build/cmsis-system_lpc11xx.o (symbol from plugin): In function `SystemInit':
(.text+0x0): multiple definition of `SystemCoreClockUpdate'
build/cmsis-system_stm32f1xx.o (symbol from plugin):(.text+0x0): first defined here
build/cmsis-system_lpc11xx.o (symbol from plugin): In function `SystemInit':
(.text+0x0): multiple definition of `SystemCoreClock'
build/cmsis-system_stm32f1xx.o (symbol from plugin):(.text+0x0): first defined here
build/cmsis-system_lpc11xx.o (symbol from plugin): In function `SystemInit':
(.text+0x0): multiple definition of `SystemInit'
build/cmsis-system_stm32f1xx.o (symbol from plugin):(.text+0x0): first defined here
collect2: error: ld returned 1 exit status
Makefile-common:93: recipe for target 'build/teacup.elf' failed

I don't know why it tries to link both lpc11xx and stm32f1xx

**Comment by Traumflug** [comment]
@Laczen, I've just added you as a collaborator, welcome to Teacup Firmware. Rules are simple: create as many branches as you feel a need for, work on branches made by others only after asking.

Here's a write up of my personal Git strategies (and a hint on a project I'm currently more active on): http://jump-ing.com/traumflugs/github_best_practices

**Comment by Laczen** [comment]
Hi, I figured out what the error was: I had to wrap my cmsis-system file in a define ``#ifdef __ARM_STM32F103__``.

**Comment by Laczen** [comment]
Hi, after further study of the next steps to take to port teacup to stm32f103 I have decided that it would take me to long to get the port done (the stm32f103 is quite different from the stm32f411) and that I wouldn't be happy with it. It would just be a messy port trying to get everything in the code.
Teacup is a nice piece of software (with a very nice config tool) but it has not been written to be easy portable (which is to be expected from something so close to the hardware).
Maybe I will try to modify teacup to use some kind of layer between hardware and Teacup (e.g. the layer that is used in klipper).

**Comment by adichell** [comment]
Hi guys,
I ported and I am porting GRBL 0.9 to some stm32 microcontroller (that are cheap and really powerful), knowing a bit of these microcontrollers, but knowing not much about GRBL. So I guess the same thing may be done or has been done with teacup somewhere, BUT there are fundamentally 2 issues to take in account:

1) Input line interrupts muxing (expecially if you want to use driver motor shields compatible with arduino)
2) Only Flash memory is available on these MCUs, there is NO EEPROM :(   This is a pain in the... brain, I solved it with a working solution, but I am still trying to find something better.

So, it's not an easy job, I found some solution to these main issues in my project, and right now I am using the ported GRBL on a Nucleo-F410 on a COREXY machine a friend built and is using. Here is my code:

[grbl_opencm3]((https://github.com/adichell/grbl_port_opencm3))

If anyone wants to port teacup that would be nice, and he may use the libopencm3, instead of using HALs or similar, and try to adopt a structure similar to mine (I have copied something of that form the libopencm3 examples project), that allows to have the logic of the firmware in a common source folder, and then compile the specific files for the different microcontrollers in separate folder.
